packages/rs-dpp/src/document/serialize.rs

```diff
@@ -1,34 +1,401 @@
+        if let Some(revision) = self.revision {
+            buffer.extend(revision.to_be_bytes())
+        }
+        document_type
+            .properties
+            .iter()
+            .try_for_each(|(field_name, field)| {
+                if field_name == CREATED_AT {
+                    if let Some(created_at) = self.created_at {
+                        buffer.extend(created_at.to_be_bytes());
+                        Ok(())
+                    } else if field.required {
+                        Err(ProtocolError::DataContractError(
+                            DataContractError::MissingRequiredKey(
+                                "created at field is not present",
+                            ),
+                        ))
+                    } else {
+                        // We don't have the created_at that wasn't required
+                        buffer.push(0);
+                        Ok(())
+                    }
+                } else if field_name == UPDATED_AT {
+                    if let Some(updated_at) = self.updated_at {
+                        buffer.extend(updated_at.to_be_bytes());
+                        Ok(())
+                    } else if field.required {
+                        Err(ProtocolError::DataContractError(
+                            DataContractError::MissingRequiredKey(
+                                "created at field is not present",
+                            ),
+                        ))
+                    } else {
+                        // We don't have the updated_at that wasn't required
+                        buffer.push(0);
+                        Ok(())
+                    }
+                } else if let Some(value) = self.properties.get(field_name) {
+                    let value = field
+                        .document_type
+                        .encode_value_ref_with_size(value, field.required)?;
+                    buffer.extend(value.as_slice());
+                    Ok(())
+                } else if field.required {
+                    Err(ProtocolError::DataContractError(
+                        DataContractError::MissingRequiredKey("a required field is not present"),
+                    ))
+                } else {
+                    // We don't have something that wasn't required
+                    buffer.push(0);
+                    Ok(())
+                }
+            })?;
+
+        Ok(buffer)
     }
 
-    pub fn serialized_size(&self) -> usize {
-        bincode::DefaultOptions::default()
-            .with_varint_encoding()
-            .reject_trailing_bytes()
-            .with_big_endian()
-            .serialized_size(self)
-            .unwrap() as usize // this should not be able to error
+    /// Serializes and consumes the document.
+    ///
+    /// The serialization of a document follows the pattern:
+    /// id 32 bytes + owner_id 32 bytes + encoded values byte arrays
+    pub fn serialize_consume(
```

Comment:
Actually the serialize_consume takes ownership of self.

Reply:
well yes, but usually it's done to prevent copy and to just move a thing out of self into output data -- zero copy, we're good

but here we're allocating a brand new vector of bytes, so in any case there will be an allocation and copy to fill it, we have no benefit of `_consume` version as far as I can see, just another portion of similar code to support and to bloat our apis